aequilibrae/project/network/gmns_exporter.py

```diff
@@ -0,0 +1,133 @@
+import pandas as pd
+from os.path import join
+from ...utils import WorkerThread
+
+from aequilibrae.parameters import Parameters
+
+
+class GMNSExporter(WorkerThread):
+    def __init__(self, net, path) -> None:
+        WorkerThread.__init__(self, None)
+        self.p = Parameters()
+        self.links = net.links
+        self.nodes = net.nodes
+        self.source = net.source
+        self.conn = net.conn
+        self.output_path = path
+
+        self.gmns_par = self.p.parameters["network"]["gmns"]
+        self.gmns_l = self.gmns_par["link"]
+        self.gmns_n = self.gmns_par["node"]
+
+    def doWork(self):
+
+        l_equiv = self.gmns_l["equivalency"]
+        n_equiv = self.gmns_n["equivalency"]
+
+        links_df = self.links.data
+        nodes_df = self.nodes.data
+
+        if "ogc_fid" in list(links_df.columns):
+            links_df.drop("ogc_fid", axis=1, inplace=True)
+        if "ogc_fid" in list(nodes_df.columns):
+            nodes_df.drop("ogc_fid", axis=1, inplace=True)
+
+        two_way_cols = list(set([col[:-3] for col in list(links_df.columns) if col[-3:] in ["_ab", "_ba"]]))
+        for idx, row in links_df.iterrows():
+
+            if row.direction == 0:
+                links_df = pd.concat([links_df, links_df.loc[idx:idx, :]], axis=0)
+                links_df.reset_index(drop=True, inplace=True)
+
+                links_df.loc[links_df.index[-1], "link_id"] = max(list(links_df.link_id)) + 1
+                links_df.loc[links_df.index[-1], "a_node"] = row.b_node
+                links_df.loc[links_df.index[-1], "b_node"] = row.a_node
+
+                links_df.loc[links_df.index[-1], "direction"] = 1
+                links_df.loc[idx, "direction"] = 1
+
+                links_df.loc[links_df.index[-1], "dir_flag"] = -1
+                links_df.loc[idx, "dir_flag"] = 1
+
+                for col in two_way_cols:
+                    links_df.loc[idx, col] = links_df.loc[idx, col + "_ab"]
+                    links_df.loc[links_df.index[-1], col] = links_df.loc[idx, col + "_ba"]
+
+            elif row.direction == -1:
+                for col in two_way_cols:
+                    links_df.loc[idx, col] = links_df.loc[idx, col + "_ba"]
+
+                links_df.loc[idx, "a_node"] = row.b_node
+                links_df.loc[idx, "b_node"] = row.a_node
+                links_df.loc[idx, "direction"] = 1
+                links_df.loc[idx, "dir_flag"] = -1
+
+            else:
+                for col in two_way_cols:
+                    links_df.loc[idx, col] = links_df.loc[idx, col + "_ab"]
+
+                links_df.loc[idx, "dir_flag"] = 1
+
+        links_df.distance = links_df.distance.apply(lambda x: x / 1000)
```

Comment:
Here we convert meters to km because, according to GMNS specifications, it's recommended to inform distances either in km or miles (long length units). However, here I'm assuming that we always have distances in meters in AequilibraE. Not sure if this assumption is always true, @pedrocamargo.
Thanks for the observation. I'll change this to be vectorised.

Reply:
You can eliminate the distance calculation. That is computed automatically by a trigger in AequilibraE